include/lsst/afw/typehandling/SimpleHeteroMap.h

```diff
@@ -0,0 +1,155 @@
+        for (auto const& keyValue : other._storage) {
+            auto const& key = keyValue.first;
+            auto const& value = keyValue.second;
+            _storage.insert(std::make_pair(key, _proxyCopy(value)));
+        }
+    }
+    SimpleHeteroMap(SimpleHeteroMap&&) = default;
+    virtual ~SimpleHeteroMap() = default;
+
+    SimpleHeteroMap& operator=(SimpleHeteroMap const& other) {
+        std::unordered_map<K, StorageType> copy;
+        for (auto const& keyValue : other._storage) {
+            auto const& key = keyValue.first;
+            auto const& value = keyValue.second;
+            copy.insert(std::make_pair(key, _proxyCopy(value)));
+        }
+        std::swap(_storage, copy);
+        return *this;
+    };
+    SimpleHeteroMap& operator=(SimpleHeteroMap&&) = default;
+
+    typename HeteroMap<K>::size_type size() const noexcept override { return _storage.size(); }
+
+    bool empty() const noexcept override { return _storage.empty(); }
+
+    typename HeteroMap<K>::size_type max_size() const noexcept override { return _storage.max_size(); }
+
+    bool contains(K const& key) const override { return _storage.count(key) > 0; }
+
+    std::vector<K> keys() const override {
+        std::vector<K> keySnapshot;
+        keySnapshot.reserve(_storage.size());
+        for (auto const& pair : _storage) {
+            keySnapshot.push_back(pair.first);
+        }
+        return keySnapshot;
+    }
+
+    void clear() noexcept override { _storage.clear(); }
+
+protected:
+    typename HeteroMap<K>::ValueReference unsafeLookup(K key) const override {
+        try {
+            return _storage.at(key);
+        } catch (std::out_of_range& e) {
+            std::stringstream message;
+            message << "Key not found: " << key;
+            std::throw_with_nested(LSST_EXCEPT(pex::exceptions::OutOfRangeError, message.str()));
+        }
+    }
+
+    bool unsafeInsert(K key, typename MutableHeteroMap<K>::InputType&& value) override {
+        return _storage.insert(std::make_pair(key, StorageType(std::move(value)))).second;
+    }
+
+    bool unsafeErase(K key) override { return _storage.erase(key); }
+
+private:
+    // Icky TMP, but I can't find another way to get at the template arguments for variant :(
+    // Method has no definition but can't be deleted without breaking definition of StorageType
+    /// @cond
+    template <typename... Types>
+    static boost::variant<std::remove_cv_t<std::remove_reference_t<Types>>...> _variadicTranslator(
+            boost::variant<Types...> const&) noexcept;
+    /// @endcond
+
+    /**
+     * The set of types that may be stored in this map.
+     *
+     * These are the modifiable value equivalents of HeteroMap<K>::ValueReference.
+     */
+    // this mouthful is shorter than the equivalent expression with result_of
+    using StorageType = decltype(_variadicTranslator(std::declval<typename HeteroMap<K>::ValueReference>()));
```

Comment:
I'll have to think about this one. The intended distinction is a bit more fundamental than "reference vs. value", in that I wanted the ideal type for a very specific role. For example, `SimpleHeteroMap::StorageType` (the value type of an internal map) and `MutableHeteroMap::InputType` (a type from which input arguments can be copied) are conceptually distinct types, even though their definitions are currently equivalent.

Reply:
I think I'll standardize the type aliases. We might lose on some subtle typing issues (e.g., should `InputType` use rvalue references instead of values?), but it will mean that almost nobody has to deal with the template metaprogramming involved in translating the variants into each other.